File: examples/ruby/oruk_client.rb
# OrukClient — minimal Net::HTTP-based client for the oruk live broadcast
# intelligence API. Standard library only.
#
#   oruk = OrukClient.new(ENV['ORUK_API_KEY'])
#   oruk.stories(category: 'conflict', min_impact: 7)['stories'].each do |s|
#     puts s['headline']
#   end

require 'net/http'
require 'uri'
require 'json'

class OrukAPIError < StandardError
  attr_reader :status, :code, :request_id
  def initialize(status, code, message, request_id = nil)
    @status, @code, @request_id = status, code, request_id
    super("oruk API: #{code} — #{message}")
  end
end

class OrukClient
  BASE_URL = 'https://api.oruk.ai'

  def initialize(api_key = nil, base_url: BASE_URL, timeout: 15)
    @api_key, @base_url, @timeout = api_key, base_url.chomp('/'), timeout
  end

  def feed(**params)    = get('/v1/stories/feed', params, requires_key: false)
  def stories(**params) = get('/v1/stories', params)
  def story(id)         = get("/v1/stories/#{URI.encode_www_form_component(id)}")
  def sources(**p)      = get('/v1/sources', p)
  def regions           = get('/v1/regions')
  def stats             = get('/v1/stats')
  def health            = get('/v1/health', {}, requires_key: false)

  # Yields every story across all pages.
  def each_page(**params)
    return to_enum(:each_page, **params) unless block_given?
    cursor = nil
    loop do
      page = stories(**(cursor ? params.merge(cursor: cursor) : params))
      (page['stories'] || []).each { |s| yield s }
      meta = page['meta'] || {}
      break unless meta['hasMore'] && meta['cursor']
      cursor = meta['cursor']
    end
  end

  private

  def get(path, params = {}, requires_key: true)
    raise "ORUK_API_KEY required for #{path}" if requires_key && @api_key.nil?

    uri = URI.parse(@base_url + path)
    uri.query = URI.encode_www_form(params.compact) unless params.empty?

    req = Net::HTTP::Get.new(uri)
    req['Accept'] = 'application/json'
    req['User-Agent'] = 'oruk-ruby-example/1.0'
    req['X-API-Key'] = @api_key if @api_key

    resp = Net::HTTP.start(uri.hostname, uri.port,
                           use_ssl: true,
                           open_timeout: 5,
                           read_timeout: @timeout) { |http| http.request(req) }

    if resp.code.to_i >= 400
      body = (JSON.parse(resp.body) rescue {})
      raise OrukAPIError.new(
        resp.code.to_i,
        body['error'] || 'http_error',
        body['message'] || "HTTP #{resp.code} from #{path}",
        resp['x-request-id'],
      )
    end
    JSON.parse(resp.body)
  end
end

File: examples/python/oruk_client.py
"""OrukClient — minimal `requests`-based client for the oruk live broadcast
intelligence API. Drop into any Python 3.9+ project.

    from oruk_client import OrukClient
    oruk = OrukClient(api_key=os.environ['ORUK_API_KEY'])
    for story in oruk.iter_pages(category='conflict', min_impact=7):
        print(story['headline'])
"""

from __future__ import annotations

import os
from typing import Any, Iterable, Iterator, Optional

import requests


class OrukAPIError(RuntimeError):
    def __init__(self, status: int, code: str, message: str, request_id: Optional[str] = None):
        super().__init__(f"oruk API: {code} — {message}")
        self.status = status
        self.code = code
        self.message = message
        self.request_id = request_id


class OrukClient:
    BASE_URL = "https://api.oruk.ai"
    DEFAULT_TIMEOUT = 15

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
        session: Optional[requests.Session] = None,
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = session or requests.Session()
        self.session.headers.update({"User-Agent": "oruk-python-example/1.0"})

    # ----- Endpoint shortcuts -----

    def feed(self, **params: Any) -> dict:
        """Public feed — works without a key."""
        return self._get("/v1/stories/feed", params=params, requires_key=False)

    def stories(self, **params: Any) -> dict:
        return self._get("/v1/stories", params=params)

    def story(self, story_id: str) -> dict:
        return self._get(f"/v1/stories/{story_id}")

    def sources(self, **params: Any) -> list:
        return self._get("/v1/sources", params=params)

    def regions(self) -> Any:
        return self._get("/v1/regions")

    def stats(self) -> dict:
        return self._get("/v1/stats")

    def health(self) -> dict:
        return self._get("/v1/health", requires_key=False)

    # ----- Pagination helper -----

    def iter_pages(self, **params: Any) -> Iterator[dict]:
        """Yield every story across as many pages as `meta.hasMore` reports."""
        cursor = None
        while True:
            page = self.stories(**(params if not cursor else {**params, "cursor": cursor}))
            for story in page.get("stories", []):
                yield story
            meta = page.get("meta", {})
            if not meta.get("hasMore") or not meta.get("cursor"):
                return
            cursor = meta["cursor"]

    # ----- Internal -----

    def _get(self, path: str, *, params: Optional[dict] = None, requires_key: bool = True) -> Any:
        headers = {"Accept": "application/json"}
        if requires_key:
            if not self.api_key:
                raise RuntimeError(f"ORUK_API_KEY is required for {path}")
        if self.api_key:
            headers["X-API-Key"] = self.api_key

        resp = self.session.get(
            self.base_url + path,
            params=params or None,
            headers=headers,
            timeout=self.timeout,
        )
        if resp.status_code >= 400:
            try:
                body = resp.json()
            except ValueError:
                body = {}
            raise OrukAPIError(
                status=resp.status_code,
                code=body.get("error", "http_error"),
                message=body.get("message", resp.text[:200] or f"HTTP {resp.status_code}"),
                request_id=resp.headers.get("x-request-id"),
            )
        return resp.json()


def from_env() -> OrukClient:
    """Convenience: build a client from ORUK_API_KEY."""
    return OrukClient(api_key=os.environ.get("ORUK_API_KEY"))
